ante-ls/src/main.rs

```diff
@@ -88,58 +91,265 @@ impl LanguageServer for Backend {
             self.update_diagnostics(params.text_document.uri, &rope).await;
         };
     }
+
+    async fn hover(&self, params: HoverParams) -> Result<Option<Hover>> {
+        self.client.log_message(MessageType::LOG, format!("ante-ls hover: {:?}", params)).await;
+        let uri = params.text_document_position_params.text_document.uri;
+        let rope = match self.document_map.get(&uri) {
+            Some(rope) => rope,
+            None => return Ok(None),
+        };
+
+        let cache = self.create_cache(&uri, &rope);
+        let ast = cache.parse_trees.get_mut(0).unwrap();
+
+        let index = position_to_index(params.text_document_position_params.position, &rope);
+        let hovered_node = walk_ast(ast, index);
+
+        match hovered_node {
+            Ast::Variable(v) => {
+                let info = match v.definition {
+                    Some(definition_id) => &cache[definition_id],
+                    _ => return Ok(None),
+                };
+
+                let typ = match &info.typ {
+                    Some(typ) => typ,
+                    None => return Ok(None),
+                };
+
+                let name = match &v.kind {
+                    VariableKind::Identifier(name) => name.clone(),
+                    VariableKind::TypeConstructor(name) => name.clone(),
+                    VariableKind::Operator(op) => format!("({})", op),
+                };
+
+                let (typ_str, traits) =
+                    typeprinter::show_type_and_traits(typ, &info.required_traits, &info.trait_info, &cache);
+
+                let value = format!("{} : {}", name, typ_str);
+                let value = if traits.is_empty() { value } else { format!("{}\n  given {}", value, traits.join(", ")) };
```

Comment:
Since this is meant to be the same as the diagnostics shown with `--show-types`, can you make a wrapper function for these lines so that this part of ante-ls and the `--show-types` flag can share the same formatting function?